### PR TITLE
[SPARK-28163][SS] Use CaseInsensitiveMap for KafkaOffsetReader

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

There are "unsafe" conversions in the Kafka connector.
`CaseInsensitiveStringMap` comes in which is then converted the following way:
```
...
options.asScala.toMap
...
```
The main problem with this is that such case it looses its case insensitive nature
(case insensitive map is converting the key to lower case when get/contains called).

In this PR I'm using `CaseInsensitiveMap` to solve this problem.

## How was this patch tested?

Existing + additional unit tests.

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
cc @HyukjinKwon and @HeartSaVioR since it's connected to SPARK-28142.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala
Comment (gaborgsomogyi) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala]: This is not a required change but thought it would be good to simplify.

**Comment by HeartSaVioR** [comment]
I feel the reason of such confusion is not explicitly requiring type to CaseInsensitiveMap whenever needed. `Map with lowercased key` and `CaseInsensitiveMap` are not same, but KafkaSourceProvider just mixes up using both. 

For the same reason, if we require lowercase key, it shouldn't be referred as "case-insensitive". It's case-sensitive, and if we can't expect the same from Kafka, we will get into trouble with passing config to Kafka.

**Comment by gaborgsomogyi** [comment]
@HeartSaVioR thanks for your comment!

This case sensitivity issue disturbs me for at least half a year so pretty open what less error prone solution can we come up. Since I've found another issue I've invested more time and analyzed it through.

Let me share my thinking and please do the same to find out how to proceed. First, start with the actual implementation. Majority of this code uses case-insensitive maps in some way (`CaseInsensitiveMap` or `CaseInsensitiveStringMap` instances arrive mainly).

> Map with lowercased key and CaseInsensitiveMap are not same

This is true.
> referring Map with lowercase key as "case-insensitive" which is not strictly true

I agree with your statement and we can discuss how to name it. Map with lowercase key is referred "case-insensitive" because the user doesn't have to care about case when the following operations used: put/get/contains. The following ops are the same:
```
val map: CaseInsensitive...
map.put("key1", "value1")
map.put("keY2", "value2")

// all is true
map.get("key1")==map.get("KEY1")
map.contains("kEy1")
map.get("key2")==map.get("KEY2")
map.contains("kEy2")
```
If the underlying conversion which is used on the keys consistently will change then the mentioned logic is still true. To find out whether user is provided a configuration I think it's the most robust solution.

The other use-case which exists is to extract the entries from these maps which start with `kafka.` and send it to Kakfa producer/consumer. This can't be called case-insensitive of course because Kafka requires lowercase keys, no question. This is the reason why the following function extracts the mentioned parameters (and force lower case conversion on keys):
```
  private def convertToSpecifiedParams(parameters: Map[String, String]): Map[String, String] = {
    parameters
      .keySet
      .filter(_.toLowerCase(Locale.ROOT).startsWith("kafka."))
      .map { k => k.drop(6).toString -> parameters(k) }
      .toMap
  }
```

Majority of this code uses the mentioned case insensitive maps but there are some parts which breaks this (for example the part which we're now modifying).

As a final conclusion from my perspective:
* Case insensitive maps has to be used for the first use-case because less error prone (this is what I tend to address)
* When any parameter passed to Kafka `convertToSpecifiedParams` must be used (this is already the situation)
* Changing the following strings to lower case would has the same effect just like the actual stand of this PR but then we stay in the same parameter case hell
```
  private[kafka010] val FETCH_OFFSET_NUM_RETRY = "fetchoffset.numretries"
  private[kafka010] val FETCH_OFFSET_RETRY_INTERVAL_MS = "fetchoffset.retryintervalms"
```

That said everything can be discussed so waiting on thoughts...

**Comment by HeartSaVioR** [comment]
First of all, thanks for the detailed analysis, @gaborgsomogyi !

I mostly agree on your analysis - just to say we are mixing up `Map with lowercased key` and `CaseInsensitiveMap`, and require `Map` as a type of parameter if we need case insensitive map so it's not forced. That was the reason of root case of #24942. We are checking the type loosely even we have proper type which ensures the requirement. That's my point.

**Comment by gatorsmile** [comment]
@HeartSaVioR @gaborfeher @dongjoon-hyun  We should not skip adding the unit test cases. Please add one. Thanks!

**Review by gatorsmile — CHANGES_REQUESTED** [review]
@HeartSaVioR @gaborfeher @dongjoon-hyun  We should not skip adding the unit test cases especially when this PR claims we are fixing a bug.. Please add one. Thanks!

**Comment by gaborgsomogyi** [comment]
@HeartSaVioR I see your point and agree the implementation can be made more safe. Since DSv1 provides only map as input parameter this can be converted to be more on the safe side and internally pass around `CaseInsensitiveMap` instances.

@gatorsmile Thanks for your review, adding test...

**Comment by gaborgsomogyi** [comment]
@HeartSaVioR I've converted `Map` to `CaseInsensitiveMap` where the mentioned bug required it to make the code less error prone. I know there are places which not yet touched but I didn't want to mix up 2 intentions, namely solve this bug and making parameter handling more robust on v1 side. Going to file another jira + PR for the second...

@gatorsmile Added tests which cover not only the problematic parts.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (dongjoon-hyun) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: Ur, could you pick up another number? This value has some religious meaning. :)

Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: Ah, didn't realize it just hit the keyboard and crafted a number which is rarely default value. Thanks for notifying it was not intentional :)

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: Why not adding them to the end of existing test, or change title of test to be specific to KafkaOffsetReader? IMHO I'm not sure it worths to have another test.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: Same here.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: Same here too.

**Review by HeartSaVioR — COMMENTED** [review]
Looks good overall. Just a 2 cents: I'm not sure we prefer renaming fields without specific reason along with other changes (I'm seeing multiple places doing it), as well as having individual tests for each JIRA issue which actually do pretty similar things.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: I've received many comments previously saying we have to keep jira in the test when it was a bugfix. In this situation I had the same feeling like you and I would personally like to have 3 tests in this case `batch`, `micro-batch` and `continuous` and forgetting about jira number. So all in all it will it will depend on the committer how it's organized.

**Comment by gaborgsomogyi** [comment]
> I'm not sure we prefer renaming fields without specific reason along with other changes

The other option was to introduce 2 different methods in the test to get the same `KafkaOffsetReader` from 2 different places (one from micro-batch and another from continuous).

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: @dongjoon-hyun WDYT, shall we have 3 tests only? Happy to modify it like that :)

Comment (dongjoon-hyun) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: Got it. In this case, +1 for that change.

Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: OK, just pushed.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala
Comment (dongjoon-hyun) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala]: Is `private val` required for `Use CaseInsensitiveMap for KafkaOffsetReader`?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala
Comment (dongjoon-hyun) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala]: - Is `private val` required for `Use CaseInsensitiveMap for KafkaOffsetReader`?
- Is the renaming from `kafkaOffsetReader` to `offsetReader` required for `Use CaseInsensitiveMap for KafkaOffsetReader`?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala
Comment (dongjoon-hyun) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala]: Is this renaming required for `Use CaseInsensitiveMap for KafkaOffsetReader`?

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
@gaborgsomogyi . I understand the intention and want to help this PR because this PR has been under reviews for 2 weeks already.

However, this PR distracts the focus inconsistently eagerly. :) In general, we should not put the refactoring and new behavior like `CaseInsensitiveMap` into a single PR. Could you revert all the irrelevant stuff? You can put them into another PR.

If the PRs become as concise as possible, we can do review and merge as fast as we can. In case of simple PRs, as you know, one or two days are enough until merging. So, which one do you want? Do you want to keep all of this in this PR? Or, do you think you can split this into the simple ones?

**Review by HeartSaVioR — APPROVED** [review]
LGTM, except still have my 2 cents for renaming fields.

**Comment by gaborgsomogyi** [comment]
@dongjoon-hyun I think the clean and long term solution is to use `CaseInsensitiveMap` inside connectors where DSv1 is involved so at some point it should be reviewed. Since we're in the middle it would be good to spend time on it. I think there are limited number of questionable parts which are already pinpointed but as always if you suggest to create here a quick fix and apply the `CaseInsensitiveMap` in another PR we can turn to this direction.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala]: Otherwise this code:
```
getField(stream, offsetReaderMethod)
```
throws exception something like:
```
java.lang.IllegalArgumentException: Can't find a private method named: offsetReader
```
The other possibility to add a getter explicitly which I thought is overkill.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala]: > Is private val required for Use CaseInsensitiveMap for KafkaOffsetReader?

Yeah, please see my previous comment. Same applies here.

> Is the renaming from kafkaOffsetReader to offsetReader required for Use CaseInsensitiveMap for KafkaOffsetReader?
 
In short, no. A little bit detailed if not renamed then the test requires something like this:
```
  private val offsetReaderMicroBatchMethod = PrivateMethod[KafkaOffsetReader]('kafkaOffsetReader)
  private val offsetReaderContinuousMethod = PrivateMethod[KafkaOffsetReader]('offsetReader)
...
    verifyFieldsInMicroBatchStream(KafkaSourceProvider.FETCH_OFFSET_NUM_RETRY, expected, stream => {
      val kafkaOffsetReader = getField(stream, offsetReaderMicroBatchMethod)
      assert(expected.toInt === getField(kafkaOffsetReader, fetchOffsetNumRetriesMethod))
    })
...
```
General considerations why I've renamed couple of variables:
* Exact same parameters are named sometimes 2-4 different ways which is hard to track
* Not renaming them would cause similar situation in the test code
* Sometimes the config name and it's internal variable has different name
* Renamed only Spark internal variables which shouldn't matter for users

We can roll back renamings, the code will work properly but I think these changes increase readability.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala]: In short, no. Similar detailed explanation applies just like in my previous comment.

Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala]: Well, the more I'm considering the renames inside `KafkaOffsetReader.scala` the more I feel it doesn't bring too much in a bugfix. It just ends-up in 2 lines change in the test so reverting them...

**Comment by dongjoon-hyun** [comment]
Okay. If this PR remains in this state, I'll leave this PR to other committers, @gaborgsomogyi and @HeartSaVioR .

**Comment by gaborgsomogyi** [comment]
@dongjoon-hyun First of all thanks for investing your time! I was not able to read your concern level properly and thought you're trying to measure the review time. Thanks for being clear in your last comment which creates different situation :)

So as I understand your main concern is this:
> we should not put the refactoring and new behavior like CaseInsensitiveMap into a single PR.

Do I get it correct that you're fine with `CaseInsensitiveMap` and renaming has to be rolled back?
The more properly I see your point the less time will it take to adopt the PR.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala]: Rolled back.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala]: The rename rolled back.
`private val` I think is essential.

**Comment by gaborgsomogyi** [comment]
@dongjoon-hyun @HeartSaVioR rolled back all the renames.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala
Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala]: Another approach would be Scala reflection but definitely more verbose than explicit getter.

**Review by HeartSaVioR — APPROVED** [review]
LGTM.

**Comment by gaborgsomogyi** [comment]
I think @dongjoon-hyun is busy with the 2.4 release so @vanzin can you have a look please?

**Comment by dongjoon-hyun** [comment]
@gaborgsomogyi . I'm just leaving this PR to the other committers. It's not busy on `2.4` release. :)

**Comment by gaborgsomogyi** [comment]
Ah, ok 🙂

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala
Comment (HyukjinKwon) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala]: We can also `private[kafka010] offsetReader: KafkaOffsetReader`. At least I found one instance:

```scala
  // Initialized when creating reader factories. If this diverges from the partitions at the latest
  // offsets, we need to reconfigure.
  // Exposed outside this object only for unit tests.
  @volatile private[sql] var knownPartitions: Set[TopicPartition] = _
```

**Comment by HyukjinKwon** [comment]
Looks fine.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousStream.scala]: Good idea, spared a couple of lines. Changed.

**Comment by gaborgsomogyi** [comment]
@HyukjinKwon thanks for investing your time!

**Comment by HeartSaVioR** [comment]
@HyukjinKwon Could we revisit this? I guess it's just waiting for last call to merge.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (HyukjinKwon) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: I know it's a bit unrelated but can we use similar way like https://github.com/apache/spark/pull/24967#discussion_r304784506 and remove those reflections? I would feel pretty said it some tests failed when I change the name of private variables ..

**Review by HyukjinKwon — APPROVED** [review]
Looks good. cc @zsxwing and @cloud-fan, I will get this in if there are no more comments in some days.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: If we would like to allow having it non-private for testing, I'll take care of this after merging. @gaborgsomogyi is on vacation (as he mentioned in other PR) so it would take couple of weeks when we would like to add the change on the scope of this PR.

Comment (HyukjinKwon) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: Can you pick up the commit and open a PR to be authored just in case some more comment come from @zsxwing and @cloud-fan.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaBatch.scala
Comment (cloud-fan) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaBatch.scala]: shall we use `CaseInsensitiveStringMap` which is used a lot in Data Source V2?

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: Actually I'm on vacation as well (I just found small amount of time to get back), so let's wait for @gaborgsomogyi to address further comments.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaBatch.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaBatch.scala]: Just back from vacation and syncing. So originally I've considered to use `CaseInsensitiveStringMap` but not chosen because of the following:
* `CaseInsensitiveStringMap` implements java `Map` and if we're passing that around then lot of useful feature would disappear (`getOrElse`, `find`, ...) or we just have to add `asScala` to all the places where scala feature are used. All in all converting forth and back was overkill from my point of view.
* `KafkaBatch` is private API and though it can be different due to the considerations above.

As always if you think it worth then I can adapt...

**Comment by cloud-fan** [comment]
retest this please

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaBatch.scala
Comment (cloud-fan) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaBatch.scala]: I think it's fine. Let's keep it as it is.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala
Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala]: Fixed.

**Comment by gaborgsomogyi** [comment]
@HyukjinKwon may I ask for another round please?

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by gaborgsomogyi** [comment]
Many thanks @cloud-fan !

**Pull request closed without merging**
